File: verman-py/project_paths.py
"""Canonical workspace metadata paths for VerMan projects."""

from __future__ import annotations

import os
from typing import Iterator, Optional


VERMAN_DIR_NAME = ".verman"
PROJECT_DB_FILENAME = "project.db"
IGNORE_FILENAME = ".vermanignore"
BACKUP_DIR_NAME = "backup"

LEGACY_DB_FILENAME = ".verman.db"
LEGACY_BACKUP_DIR_NAME = ".verman_backup"
DATABASE_SIDE_SUFFIXES = ("", "-wal", "-shm", "-journal")


def normalize_workspace_path(workspace_path: str) -> str:
    return os.path.abspath(workspace_path)


def get_metadata_dir(workspace_path: str) -> str:
    return os.path.join(normalize_workspace_path(workspace_path), VERMAN_DIR_NAME)


def ensure_metadata_dir(workspace_path: str) -> str:
    metadata_dir = get_metadata_dir(workspace_path)
    os.makedirs(metadata_dir, exist_ok=True)
    return metadata_dir


def get_project_database_path(workspace_path: str) -> str:
    return os.path.join(get_metadata_dir(workspace_path), PROJECT_DB_FILENAME)


def get_ignore_file_path(workspace_path: str) -> str:
    return os.path.join(normalize_workspace_path(workspace_path), IGNORE_FILENAME)


def get_backup_dir(workspace_path: str) -> str:
    return os.path.join(get_metadata_dir(workspace_path), BACKUP_DIR_NAME)


def get_legacy_database_path(workspace_path: str) -> str:
    return os.path.join(normalize_workspace_path(workspace_path), LEGACY_DB_FILENAME)


def get_legacy_backup_dir(workspace_path: str) -> str:
    return os.path.join(normalize_workspace_path(workspace_path), LEGACY_BACKUP_DIR_NAME)


def iter_database_sidecar_paths(database_path: str) -> Iterator[str]:
    for suffix in DATABASE_SIDE_SUFFIXES:
        yield f"{database_path}{suffix}"


def find_existing_database_path(workspace_path: str) -> Optional[str]:
    database_path = get_project_database_path(workspace_path)
    if os.path.exists(database_path):
        return database_path

    legacy_database_path = get_legacy_database_path(workspace_path)
    if os.path.exists(legacy_database_path):
        return legacy_database_path

    return None


def is_project_workspace(workspace_path: str) -> bool:
    return find_existing_database_path(workspace_path) is not None

File: verman-rust/src-tauri/src/project_paths.rs
use std::path::{Path, PathBuf};

pub const VERMAN_DIR_NAME: &str = ".verman";
pub const PROJECT_DB_FILENAME: &str = "project.db";
pub const IGNORE_FILENAME: &str = ".vermanignore";
pub const BACKUP_DIR_NAME: &str = "backup";
pub const LEGACY_DB_FILENAME: &str = ".verman.db";
pub const LEGACY_BACKUP_DIR_NAME: &str = ".verman_backup";
pub const DATABASE_SIDE_SUFFIXES: &[&str] = &["", "-wal", "-shm", "-journal"];

pub fn get_metadata_dir(workspace_path: &Path) -> PathBuf {
    workspace_path.join(VERMAN_DIR_NAME)
}

pub fn ensure_metadata_dir(workspace_path: &Path) -> std::io::Result<PathBuf> {
    let dir = get_metadata_dir(workspace_path);
    std::fs::create_dir_all(&dir)?;
    Ok(dir)
}

pub fn get_project_database_path(workspace_path: &Path) -> PathBuf {
    get_metadata_dir(workspace_path).join(PROJECT_DB_FILENAME)
}

pub fn get_ignore_file_path(workspace_path: &Path) -> PathBuf {
    workspace_path.join(IGNORE_FILENAME)
}

pub fn get_backup_dir(workspace_path: &Path) -> PathBuf {
    get_metadata_dir(workspace_path).join(BACKUP_DIR_NAME)
}

pub fn get_legacy_database_path(workspace_path: &Path) -> PathBuf {
    workspace_path.join(LEGACY_DB_FILENAME)
}

pub fn get_legacy_backup_dir(workspace_path: &Path) -> PathBuf {
    workspace_path.join(LEGACY_BACKUP_DIR_NAME)
}

pub fn iter_database_sidecar_paths(db_path: &Path) -> Vec<PathBuf> {
    let base = db_path.to_string_lossy().to_string();
    DATABASE_SIDE_SUFFIXES
        .iter()
        .map(|suffix| PathBuf::from(format!("{}{}", base, suffix)))
        .collect()
}

pub fn find_existing_database_path(workspace_path: &Path) -> Option<PathBuf> {
    let db_path = get_project_database_path(workspace_path);
    if db_path.exists() {
        return Some(db_path);
    }
    let legacy_db_path = get_legacy_database_path(workspace_path);
    if legacy_db_path.exists() {
        return Some(legacy_db_path);
    }
    None
}

pub fn get_hash_cache_path(workspace_path: &Path) -> PathBuf {
    get_metadata_dir(workspace_path).join("hash_cache.json")
}

pub fn get_blobs_dir(workspace_path: &Path) -> PathBuf {
    get_metadata_dir(workspace_path).join("blobs")
}

pub fn get_blob_path(workspace_path: &Path, file_hash: &str) -> PathBuf {
    get_blobs_dir(workspace_path).join(file_hash)
}

pub fn is_project_workspace(workspace_path: &Path) -> bool {
    find_existing_database_path(workspace_path).is_some()
}
